pkg/sql/sessionprotectedts/session_protected_ts_test.go

```diff
@@ -0,0 +1,162 @@
+	targetDB := getUserConn(t, s0)
+	sqlutils.MakeSQLRunner(targetDB).Exec(t, `SELECT 1`)
+
+	// Fetch the IDLE session's ID.
+	var idleSessionID string
+	sqlDB.QueryRow(t, `SELECT session_id FROM crdb_internal.cluster_sessions WHERE last_active_query LIKE 'SELECT 1%' ORDER BY session_start DESC LIMIT 1`).Scan(&idleSessionID)
+	recordID := uuid.MakeV4()
+	rec := sessionprotectedts.MakeRecord(recordID, []byte(idleSessionID), s0.Clock().Now(), ptpb.MakeClusterTarget())
+	require.NoError(t, pts.Protect(ctx, rec))
+
+	return targetDB, rec
+}
+
+func TestSessionProtectedTimestampReconciler(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	ctx := context.Background()
+	defer log.Scope(t).Close(t)
+
+	testCluster := serverutils.StartCluster(t, 1, base.TestClusterArgs{})
+	defer testCluster.Stopper().Stop(ctx)
+	s0 := testCluster.Server(0).ApplicationLayer()
+
+	sqlDB := sqlutils.MakeSQLRunner(s0.SQLConn(t, "defaultdb"))
+	_ = sqlDB.Exec(t, `
+CREATE USER foo with password 'hunter2';
+`)
+
+	insqlDB := s0.InternalDB().(isql.DB)
+	ptp := s0.ExecutorConfig().(sql.ExecutorConfig).ProtectedTimestampProvider
+	pts := ptstorage.WithDatabase(ptp, insqlDB)
+
+	settings := cluster.MakeTestingClusterSettings()
+	r := ptreconcile.New(settings, insqlDB, ptp,
+		ptreconcile.StatusFuncs{
+			sessionprotectedts.SessionMetaType: sessionprotectedts.MakeStatusFunc(),
+		})
+	require.NoError(t, r.StartReconciler(ctx, s0.AppStopper()))
+
+	// Create two sessions, these will be IDLE as they are running SELECT queries.
+	s1, rec1 := createSessionAndProtect(ctx, t, s0, sqlDB, pts)
+	s2, rec2 := createSessionAndProtect(ctx, t, s0, sqlDB, pts)
+
+	state, err := pts.GetState(ctx)
+	require.NoError(t, err)
+	require.Equal(t, uint64(2), state.NumRecords)
+
+	t.Run("reconcile", func(t *testing.T) {
+		ptreconcile.ReconcileInterval.Override(ctx, &settings.SV, time.Millisecond)
+		// Close one of the sessions, and ensure that the reconciler removes the
+		// associated PTS record.
+		require.NoError(t, s1.Close())
+		testutils.SucceedsSoon(t, func() error {
+			require.Equal(t, int64(0), r.Metrics().ReconciliationErrors.Count())
+			if removed := r.Metrics().RecordsRemoved.Count(); removed != 1 {
+				return errors.Newf("expected processed to be 1, got %d", removed)
+			}
+			return nil
+		})
+		_, err := pts.GetRecord(ctx, rec1.ID.GetUUID())
+		require.Regexp(t, protectedts.ErrNotExists, err)
+
+		// Verify the second record is still around.
+		_, err = pts.GetRecord(ctx, rec2.ID.GetUUID())
+		require.NoError(t, err)
+		require.NoError(t, s2.Close())
+
+		// Open one last session and keep it active.
+		s3, rec3 := createSessionAndProtect(ctx, t, s0, sqlDB, pts)
+		defer func() {
+			_ = s3.Close()
+		}()
+		var wg sync.WaitGroup
+		wg.Add(1)
+		aboutToSleep := make(chan struct{})
+		defer close(aboutToSleep)
+		go func() {
+			defer wg.Done()
+			aboutToSleep <- struct{}{}
+			_, err := s3.Exec(`SELECT pg_sleep(3)`)
+			require.NoError(t, err)
+		}()
+		<-aboutToSleep
```

Comment:
I suppose this is still racy, but I don't think that it'll cause a test failure.

Reply:
yeah i did consider putting a longer sleep to increase the chances of us catching an "active" query but it makes the test unnecessarily long. I'm hopeful a few runs will atleast be active instead of idle and give us coverage.